sparse/coo/common.py

```diff
@@ -609,3 +609,58 @@ def nanreduce(x, method, identity=None, axis=None, keepdims=False, **kwargs):
     """
     arr = _replace_nan(x, method.identity if identity is None else identity)
     return arr.reduce(method, axis, keepdims, **kwargs)
+
+
+def roll(a, shift, axis=None):
+    """
+    Shifts elements of an array along specified axis. Elements that roll beyond
+    the last position are circulated and re-introduced at the first.
+
+    Parameters
+    ----------
+    x : COO
+        Input array
+    shift : int or tuple of ints
+        Number of index positions that elements are shifted. If a tuple is
+        provided, then axis must be a tuple of the same size, and each of the
+        given axes is shifted by the corresponding number (duplicate entries
+        for each axis are allowed). If an int while axis is a tuple of ints,
+        then broadcasting is used so the same shift is applied to all axes.
+    axis : int or tuple of ints, optional
+        Axis or tuple specifying multiple axes. By default, the
+        array is flattened before shifting, after which the original shape is
+        restored.
+
+    Returns
+    -------
+    res : ndarray
+        Output array, with the same shape as a.
+    """
+    from numpy.core.numeric import normalize_axis_tuple
+    from .core import COO
+
+    # rolling does nothing for empty array
+    if a.nnz == 0:
+        return COO(a.coords, data=a.data, shape=a.shape)
+
+    # roll flattened array
+    elif axis is None:
+        return roll(a.reshape((-1,)), shift, 0).reshape(a.shape)
+
+    # roll across specified axis
+    else:
+        axis = normalize_axis_tuple(axis, a.ndim, allow_duplicate=True)
+        broadcasted = np.core.multiarray.broadcast(shift, axis)
```

Comment:
Is this just broadcasting the shape? If so, we should use our own utility function. See https://github.com/pydata/sparse/blob/b51d74924d62ff6537b15ce4e1dd4e56080a3b6f/sparse/coo/umath.py#L358-L388

In general, I don't think it's good to rely on an undocumented NumPy function, which could change and break things at any time.

Reply:
Hmm what about [`normalize_axis_tuple`](https://github.com/numpy/numpy/blob/59bd9ee273f26f335f1f77e0b85ede8d209a7e07/numpy/core/numeric.py#L1469) is there a utility function for that as well?

Reply:
Nevermind I found `normalize_axis` in utils.

Reply:
You'll need to add `allow_repeat` into that one, I believe.